url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/commands/model3/mt/MTModule.java

```diff
@@ -0,0 +1,583 @@
+import generativemodel.ModelTrackedVarIterator;
+import generativemodel.RVariable;
+import generativemodel.RVariableValue;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+
+import burlap.debugtools.RandomFactory;
+import burlap.oomdp.core.GroundedProp;
+
+import commands.model3.StringValue;
+import commands.model3.TaskModule.LiftedVarValue;
+
+public class MTModule extends GMModule {
+
+	public static final String						SNAME = "semanticCommand";
+	public static final String						NNAME = "naturalCommand";
+	
+	
+	protected RVariable								liftedRFVariable;
+	protected RVariable								bindingConstraintVariable;
+	
+	protected RVariable								semanticCommandVariable;
+	protected RVariable								naturalCommandVariable;
+	
+	
+	protected Set<String> 							semanticWords;
+	protected Set<String> 							naturalWords;
+	
+	protected int									maxSemanticCommandLength;
+	protected int									maxNaturalCommandLength;
+	
+	protected DistortionParam 						dp;
+	protected WordParam 							wp;
+	protected LengthParam 							lp;
+	
+	protected Tokenizer								naturalTokenizer;
+	protected Tokenizer								semanticTokenizer;
+	
+	protected Random								rand;
+	
+	public MTModule(String name, RVariable liftedRFVariable, RVariable bindingConstraintVariable, Set<String> semanticWords, Set<String> naturalWords, 
+			int maxSemanticCommandLength, int maxNaturalCommandLenth, Tokenizer tokenizer) {
+		super(name);
+		
+		this.liftedRFVariable = liftedRFVariable;
+		this.bindingConstraintVariable = bindingConstraintVariable;
+		
+		this.semanticCommandVariable = new RVariable(SNAME, this);
+		this.naturalCommandVariable = new RVariable(NNAME, this);
+		
+		this.semanticWords = semanticWords;
+		this.naturalWords = naturalWords;
+		
+		this.maxSemanticCommandLength = maxSemanticCommandLength;
+		this.maxNaturalCommandLength = maxNaturalCommandLenth;
+		
+		this.naturalTokenizer = tokenizer;
+		this.semanticTokenizer = new Tokenizer(true);
+		
+		this.wp = new WordParam();
+		this.dp = new DistortionParam();
+		this.lp = new LengthParam();
+		
+		//initialize the word parameters to be uniform
+		double uniWordGen = 1. / this.semanticWords.size();
+		for(String s : this.semanticWords){
+			for(String n : this.naturalWords){
+				this.wp.set(uniWordGen, n, s);
+			}
+		}
+		
+		//initialize length params; this should probably be reset once a dataset is provided
+		double uniLen = 1. / this.maxNaturalCommandLength;
+		for(int l = 1; l <= this.maxSemanticCommandLength; l++){
+			for(int m = 1; m <= this.maxNaturalCommandLength; m++){
+				this.lp.set(uniLen, l, m);
+				
+				for(int i = 1; i <= m; i++){
+					double uniDist = 1./(l+1); //old form, which seems like it was wrong: 1./(this.maxL+1);
+					for(int j = 0; j <= l; j++){
+						this.dp.set(uniDist, j, i, l, m);
+					}
+				}
+				
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 85:97, 125:137.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.